# As we see in (0EDA Notebook) in Google_playstote dataset there is some Missing data in Five Columns so in this notebook we going to handle those Missing values with different-different imputation technique.

# These are a Five Column who have an Missing data:-

    (1) Ratings 
    
    (2) Type
    
    (3) Content Rating
    
    (4) Current Ver
    
    (5) Android Ver

## Importing all necessary modules

In [2]:
import pandas as pd

In [3]:
# Loading dataset
df_google = pd.read_csv('googleplaystore.csv')

In [4]:
df_google

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

### Extracting only those columns who have a missing data with a number of missing data in respect to each column

In [5]:
missing_data_col = [(i,df_google.shape[0]-df_google[i].count()) for i in df_google.columns if df_google[i].count()<df_google.shape[0]]

In [6]:
missing_data_col

[('Rating', 1474),
 ('Type', 1),
 ('Content Rating', 1),
 ('Current Ver', 8),
 ('Android Ver', 3)]

In [7]:
df_google.isnull().mean().sort_values(ascending=False)[0:5]

Rating            0.135965
Current Ver       0.000738
Android Ver       0.000277
Type              0.000092
Content Rating    0.000092
dtype: float64

## 1st Technique:-- Removing observation with missing data.

### Note:-- This is mostly used when the data is missing randomly and only small portion of the data is missing if number of missing data is high then it leads us to removal of a big portion of a dataset which is not good.

#### As we see in four columns ('Type', 'Content Rating', 'Current Ver', 'Android Ver') very very few missing values so these technique will we good for there four column but not for Ratings columns

#### Droping a records where ('Type', 'Content Rating', 'Current Ver', 'Android Ver') have a null data with the help of drop function.

In [8]:
df_google.dropna(subset=['Type', 'Content Rating', 'Current Ver', 'Android Ver'], inplace=True)

In [9]:
df_google.isnull().sum()

App                  0
Category             0
Rating            1469
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          0
Android Ver          0
dtype: int64

## Now you see that there is only one column left with missing data which is Ratings.

## After dropping all of those records size of data is also reduce.

In [10]:
df_google.shape

(10829, 13)

## Conclusion:- Previously it was (10841, 13) and now it is (10829, 13) that means 12 records are dropped from our dataset.

## 2nd Technique:-- Doing mean and median imputation.

## Note:-- This imputation technique is only works with the numerical variables be can't use this with categorical variables this mostly usefull technique when large portion of data is missing from the columns

## In this techique we replace the null/missing values with the mean or the median of the variables.


## As we know in Ratings columns we have a outliers and mean is always heighly affected bt the outliers so here we can go with a median.

#### Replacing missing values of Ratings columns with its median or central value

In [11]:
df_google.Rating.isnull().sum()

1469

In [12]:
ratings_median = df_google['Rating'].median()

In [13]:
ratings_median

4.3

In [14]:
df_google['Rating'].fillna(ratings_median, inplace=True)

In [15]:
df_google.Rating.isnull().sum()

0

#### Now we can see that there is no missing values in this column because be replaced it with the median value of the Ratings.

In [16]:
df_google.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

# So in this we handled all the missing data so in this dataset didn't have any null or missing data.

In [17]:
df_google.to_csv('gps_missing-value-handled-dataset.csv', index=False)